# Lab 3: Exploratory Data Analysis with Statistical Testing

**Daniel Leskiewicz**

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from scipy import stats

## Problem Results

You now decide to look at the survey results. You have 100 responses! 68 students said
they play video games regularly, while 32 students said they did not. The 68 games
have an average GPA of 3.4 with a standard deviation of 1.2, while the 32 non-gamers
have an average GPA of 3.3 with a standard deviation of 1.1.

## Part 1: Review of Statistical Tests

##### We use a t-test when we have one nominal variable and one measurement variable, and we want to compare the mean values of the measurement variable. The situtation of the problem meets the criteria because we have a nomial variable which is the Yes or No question asking Do you play video games regularly and the measurement variable is the GPA. 

##### The null hypothesis is that there is no difference between if the person plays games or not and their GPA. The alternate hypothesis is if there is a difference between a person playing and their GPA.

##### We would use the two smaple t test when we have determined if two population means are equal. This is typically used with numerical and or boolean values.

In [2]:
stats.ttest_ind_from_stats(mean1=3.4, std1=1.2, nobs1=68, mean2=3.3, std2=1.1, nobs2=32)

Ttest_indResult(statistic=0.39893881176878243, pvalue=0.6908062583072547)

##### From the results above, the pvalue is 0.69080 which is greater than 0.01 which then concludes that we are unable to reject the null hypothesis because the pvalue is greater than the threshold given. The difference is not significant between the groups as a result.

##### Well the evidence is that it is supported by the original null hypothesis that students that play videos games their GPA's are not affected negatively.

## Part 2: Exploring Additional Statistical Tests

Using three different statistical tests and answering questions on these tests.

a)  List the two types of variables for which test is appropiate.


b)  Write down the general forms of the null and alternate hypotheses (one sentence per)


c)  In your own words, write what it would mean if the test did and did not indicate statistical significance. 


### Correlation Based on Linear Regression

a) Correlation Based on Linear Regression uses two numerical values

b) 

Null Hypothesis: Is the slope of the best-fit line equal to zero
   
Alternative Hypothesis: Is the slope of the best-fit line not equal to zero

c) So to start, if they are significant then they correlated otherwise they are not. 

## Kruskal-Wallis Test

a) The Kruskal-Wallis test uses one nominal variables and one ranked variable. This tests shows whether the mean ranks are the same in all the groups.

b) 

Null Hypothesis: The test is that the mean ranks of the groups are the same.

Alternative Hypothesis: The tests is that the mean ranks of the groups are not the same.

c) So to start, if the test indicated statistical significance then the two variables are associated together or wise they are not.

## Chi-Sqaured Test of Goodness of fit

a) Chi-sqaured test of goodness is used with one nominal variable, you want to see whether the number of observations in each category fits a theoretical expectation, and the sample size if large.

b) 

Null Hypothesis: The number of observations in each category that is equal to the predicted by a biological theory.

Alternative Hypothesis: Is the observed numbers are different from the expected.

c) 

To start, you take your calculated chi-sqaured value and compare it to a critical value from the chi-square table, if the chi-sqaure value is more than the critical value, then there is a significant difference.

## Part 3: Regression on Price

#### a) Load the CSV file of the cleaned dataset 

In [3]:
data = pd.read_csv('UpdatedAndCleanedSacramentorealestatetransactions.csv', dtype={'city': 'category', 'zip': 'category', \
                      'state': 'category', 'beds': 'category', \
                      'baths': 'category', 'type': 'category', \
                      'street_type': 'category'})

In [4]:
data.head(10)

street            city    zip state beds baths  \
0                     3526 HIGH ST      SACRAMENTO  95838    CA    2     1   
1                      51 OMAHA CT      SACRAMENTO  95823    CA    3     1   
2                   2796 BRANCH ST      SACRAMENTO  95815    CA    2     1   
3                 2805 JANETTE WAY      SACRAMENTO  95815    CA    2     1   
4                  6001 MCMAHON DR      SACRAMENTO  95824    CA    2     1   
5               5828 PEPPERMILL CT      SACRAMENTO  95841    CA    3     1   
6              6048 OGDEN NASH WAY      SACRAMENTO  95842    CA    3     2   
7                    2561 19TH AVE      SACRAMENTO  95820    CA    3     1   
8  11150 TRINITY RIVER DR Unit 114  RANCHO CORDOVA  95670    CA    2     2   
9                     7325 10TH ST       RIO LINDA  95673    CA    3     2   

   sq__ft         type                     sale_date  price   latitude  \
0     836  Residential  Wed May 21 00:00:00 EDT 2008  59222  38.631913   
1    1167  Residential  Wed May 21 00:00:00 EDT 2008  68212  38.478902   
2     796  Residential  Wed May 21 00:00:00 EDT 2008  68880  38.618305   
3     852  Residential  Wed May 21 00:00:00 EDT 2008  69307  38.616835   
4     797  Residential  Wed May 21 00:00:00 EDT 2008  81900  38.519470   
5    1122        Condo  Wed May 21 00:00:00 EDT 2008  89921  38.662595   
6    1104  Residential  Wed May 21 00:00:00 EDT 2008  90895  38.681659   
7    1177  Residential  Wed May 21 00:00:00 EDT 2008  91002  38.535092   
8     941        Condo  Wed May 21 00:00:00 EDT 2008  94905  38.621188   
9    1146  Residential  Wed May 21 00:00:00 EDT 2008  98937  38.700909   

    longitude  empty_lot street_type  
0 -121.434879      False          ST  
1 -121.431028      False          CT  
2 -121.443839      False          ST  
3 -121.439146      False         WAY  
4 -121.435768      False          DR  
5 -121.327813      False          CT  
6 -121.351705      False         WAY  
7 -121.481367      False         AVE  
8 -121.270555      False        Unit  
9 -121.442979      False          ST

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 985 entries, 0 to 984
Data columns (total 14 columns):
street         985 non-null object
city           985 non-null category
zip            985 non-null category
state          985 non-null category
beds           985 non-null category
baths          985 non-null category
sq__ft         985 non-null int64
type           985 non-null category
sale_date      985 non-null object
price          985 non-null int64
latitude       985 non-null float64
longitude      985 non-null float64
empty_lot      985 non-null bool
street_type    985 non-null category
dtypes: bool(1), category(7), float64(2), int64(2), object(2)
memory usage: 60.1+ KB


#### b) Fitting a simple linear regression model

In [6]:
slope, intercept, r, sq__ft_pvalue, stderr = stats.linregress(data['price'], data['sq__ft'])
slope, intercept, r, lat_pvalue, stderr = stats.linregress(data['price'], data['latitude'])
slope, intercept, r, long_pvalue, stderr = stats.linregress(data['price'], data['longitude'])
print("Pvalue for Sq__ft: ", sq__ft_pvalue)
print("Pvalue for latitude: ", lat_pvalue)
print("Pvalue for longitude: ", long_pvalue)

Pvalue for Sq__ft:  4.433056844561304e-27
Pvalue for latitude:  0.21464106576976388
Pvalue for longitude:  8.552356644182237e-20


| Variable | Pvalue | Significance at Threshold of 0.01
| --- | --- | --- |
| Sq Ft | 4.43305e-27 | Yes |
| Latitude | .214641 | No |
| Longitude | 8.55235e-20 | Yes |

#### c) We can test for association between categorical and continuous variables using Kruskal-Wallis test

In [7]:
city_samples = []
for v in set(data['city']):
    mask = data['city'] == v
    city_samples.append(data['price'][mask])
value, city_pvalue = stats.kruskal(*city_samples)
print(city_pvalue)

8.588143361577209e-49


In [9]:
samples_by_zip = []
for v in set(data['zip']):
    mask = data['zip'] == v
    samples_by_zip.append(data['price'][mask])
value2, zip_p = stats.kruskal(*samples_by_zip)
print(zip_p)

2.2206936511400606e-65


In [10]:
samples_by_beds = []
for value in set(data['beds']):
    mask = data['beds'] == value
    samples_by_beds.append(data['price'][mask])
value3, beds_p = stats.kruskal(*samples_by_beds)
print(beds_p)

9.323012566322206e-38


In [11]:
samples_by_baths = []
for value in set(data['baths']):
    mask = data['baths'] == value
    samples_by_baths.append(data['price'][mask])
value4, baths_p = stats.kruskal(*samples_by_baths)
print(baths_p)

1.1109110373679834e-50


In [12]:
samples_by_type = []
for value in set(data['type']):
    mask = data['type'] == value
    samples_by_type.append(data['price'][mask])
value5, type_p = stats.kruskal(*samples_by_type)
print(type_p)

1.1531292565462277e-06


In [13]:
samples_by_empty_lot = []
for value in set(data['empty_lot']):
    mask = data['empty_lot'] == value
    samples_by_empty_lot.append(data['price'][mask])
value6, empty_lot_p = stats.kruskal(*samples_by_empty_lot)
print(empty_lot_p)

0.03803941698702659


In [14]:
samples_by_street_type = []
for value in set(data['street_type']):
    mask = data['street_type'] == value
    samples_by_street_type.append(data['price'][mask])
value7, street_type_p = stats.kruskal(*samples_by_street_type)
print(street_type_p)

4.552769035566837e-22


| Variable | p-value | Statistically Significant Relationship at Threshold = 0.01 |
| --- | --- | --- |
| City | 8.588143361571794e-49 | Yes |
| Zip | 2.2206936511391717e-65 | Yes |
| State | NA | No |
| Beds | 9.323012566322206e-38 | Yes |
| Baths | 1.1109110373679834e-50 | Yes |
| Type | 1.1531292565464816e-06 | Yes |
| Empty_Lot | 0.03803941698702659 | No |
| Street_Type | 4.552769035566837e-22 | Yes |

d) How do the results of the statistical tests compare with your analysis of the visualizations of variable pairs from Lab 2?
- Well to start the results of the statistical tests are mostly up to par with the analysis I made from lab 2. I would say that the state could have more statisical significance if this data was across multiple states. But for the most part all of the significance are represeneted by the city, zip, amount of baths, amount of beds, but within lab 2 I had that street type was not a good predictor upon price meanwhile using the Kruskal test we see that it has some significance.

## Part 4: Classification on Property Type

#### a) Running the Kruskal-Wallis Test for the continuous variables versus the property type

In [15]:
samples_by_type = []
for value in set(data['type']):
    mask = data['type'] == value
    samples_by_type.append(data['sq__ft'][mask])
value8, sq_ft_p = stats.kruskal(*samples_by_type)
print(sq_ft_p)

3.849635133463987e-12


In [16]:
samples_by_type = []
for value in set(data['type']):
    mask = data['type'] == value
    samples_by_type.append(data['price'][mask])
value9, price_p = stats.kruskal(*samples_by_type)
print(price_p)

1.1531292565462277e-06


In [17]:
samples_by_type = []
for value in set(data['type']):
    mask = data['type'] == value
    samples_by_type.append(data['latitude'][mask])
value10, latitude_p = stats.kruskal(*samples_by_type)
print(latitude_p)

0.4933696138791226


In [18]:
samples_by_type = []
for value in set(data['type']):
    mask = data['type'] == value
    samples_by_type.append(data['longitude'][mask])
value11, longitude_p = stats.kruskal(*samples_by_type)
print(longitude_p)

0.3574295072205312


| Variable | p-value | Statistically Significant Relationship at Threshold = 0.01 |
| --- | --- | --- |
| Sq Ft | 3.849635133463987e-12 | Yes |
| Price | 31.1531292565464816e-06 | Yes |
| Latitude | 0.493369613879038 | No |
| Longitude | 0.3574295072205312 | No |

#### b) Testing two categorical variables for association using a Chi-squared test of independence.

In [19]:
combination_counts = data[['type', 'price']]\
                    .groupby(by=['type', 'price'])\
                    .size().unstack(level=0).fillna(0)
chi2, price_pValue, _, _ = stats.chi2_contingency(combination_counts)
print(price_pValue)

0.9999997357037187


In [20]:
combination_counts = data[['type', 'city']]\
                    .groupby(by=['type', 'city'])\
                    .size().unstack(level=0).fillna(0)
chi2, city_pValue, _, _ = stats.chi2_contingency(combination_counts)
print(city_pValue)

6.884142190454073e-149


In [21]:
combination_counts = data[['type', 'zip']]\
                    .groupby(by=['type', 'zip'])\
                    .size().unstack(level=0).fillna(0)
chi2, zip_pValue, _, _ = stats.chi2_contingency(combination_counts)
print(zip_pValue)

6.195871832483889e-06


In [22]:
combination_counts = data[['type', 'state']]\
                    .groupby(by=['type', 'state'])\
                    .size().unstack(level=0).fillna(0)
chi2, state_pValue, _, _ = stats.chi2_contingency(combination_counts)
print(state_pValue)

1.0


In [23]:
combination_counts = data[['type', 'beds']]\
                    .groupby(by=['type', 'beds'])\
                    .size().unstack(level=0).fillna(0)
chi2, beds_pValue, _, _ = stats.chi2_contingency(combination_counts)
print(beds_pValue)

1.6090894896930032e-64


In [24]:
combination_counts = data[['type', 'baths']]\
                    .groupby(by=['type', 'baths'])\
                    .size().unstack(level=0).fillna(0)
chi2, baths_pValue, _, _ = stats.chi2_contingency(combination_counts)
print(baths_pValue)

2.225837221388893e-41


In [25]:
combination_counts = data[['type', 'empty_lot']]\
                    .groupby(by=['type', 'empty_lot'])\
                    .size().unstack(level=0).fillna(0)
chi2, empty_lot_pValue, _, _ = stats.chi2_contingency(combination_counts)
print(empty_lot_pValue)

0.03857273017762252


In [26]:
combination_counts = data[['type', 'street_type']]\
                    .groupby(by=['type', 'street_type'])\
                    .size().unstack(level=0).fillna(0)
chi2, street_type_pValue, _, _ = stats.chi2_contingency(combination_counts)
print(street_type_pValue)

3.417489578166013e-79


| Variable | p-value | Statistically Significant Relationship at Threshold = 0.01 |
| --- | --- | --- |
| Price | 0.9999997357037187 | No |
| City | 6.884142190454073e-149 | No |
| Zip | 6.195871832483889e-06 | Yes |
| State | 1.0 | No |
| Beds | 1.6090894896930032e-64 | Yes |
| Baths | 2.225837221388893e-41 | Yes |
| Empty_Lot | 0.03857273017762252 | No |
| Street_Type | 3.417489578166013e-79 | Yes |

c) How do the results of the statistical result compare with your analysis of the visulations of variables pairs from Lab 2?
 - Overall, we can see that comparing to lab 2, we see that within lab 2 the significance was street type, price, baths, beds, sqft, city, and zip. Meanwhile using the Chi-Sqaured test we can see that the signifcant variables were street type, baths, beds, and zip. The difference here is that City is not included as of one significant variable when using the Chi-Sqaured test. The association from the heatmaps weren't that clear for me so I was incorrect on stating that it was significant at that time of peforming and analyzing lab 2.